### ISSUE TITLE
unable to segment using U-Net architecture

### ISSUE THREAD

**Issue by sudhasubramaniam** [tensorflow/hub#241]
I tried to segment my image using u-net architecture, the accuracy is very less around 10% and moreover I'm unable to print the segmented output. Attached code for your reference.

import matplotlib.pyplot as plt
import tensorflow as tf
import pandas as pd
import numpy as np
from sklearn import metrics
from sklearn.metrics import confusion_matrix
import time
from datetime import timedelta
import math
import dataset1
import random
import tensorlayer as t1
from tensorlayer.layers import *

import numpy as np,sys,os
from sklearn.utils import shuffle
from scipy.ndimage import imread
from scipy.misc import imresize

    

# Number of color channels for the images: 1 channel for gray-scale.
num_channels = 3

# image dimensions (only squares for now)
img_size = 600
img_size1=200

# Size of image when flattened to a single dimension
img_size_flat = img_size1 * img_size * num_channels

# Tuple with height and width of images used to reshape arrays.
img_shape = (img_size1, img_size)

# class info

# batch size
batch_size = 40

# validation split
validation_size = .4

# how long to wait after validation loss stops improving before terminating training
early_stopping = None  # use None if you don't want to implement early stoping

plt_epoch=[]
plt_loss=[]
plt_acc=[]
fpr1=[]
tpr1=[]
train_path='training data'
#test_path='testing data/images'
labels='training data'

data = dataset1.read_train_sets(train_path, img_size1,img_size, labels, validation_size=validation_size)
#test_images, test_ids = dataset1.read_test_set(test_path, img_size1,img_size)

print("Size of:")
print("- Training-set:\t\t{}".format(len(data.train.labels)))
#print("- Test-set:\t\t{}".format(len(test_images)))
print("- Validation-set:\t{}".format(len(data.valid.labels)))

def new_weights(shape):
    return tf.Variable(tf.truncated_normal(shape, stddev=0.05))

def new_biases(length):
    return tf.Variable(tf.constant(0.05, shape=[length]))

session = tf.Session()

x = tf.placeholder(shape=[None,200,600,3],dtype=tf.float32,name='x')
y_true = tf.placeholder(shape=[None,200,600,3],dtype=tf.float32,name='y_true')
y_true_cls = tf.argmax(y_true, dimension=1)
#y_true_cls = tf.argmax(y_true, dimension=1)

keep_prob = tf.placeholder(tf.float32) 

#keep_prob = tf.placeholder(tf.float32) # dropout (keep probability)

np.random.seed(678)
tf.set_random_seed(5678)

def tf_relu(x): return tf.nn.relu(x)
def d_tf_relu(s): return tf.cast(tf.greater(s,0),dtype=tf.float32)
def tf_softmax(x): return tf.nn.softmax(x)
def np_sigmoid(x): 1/(1 + np.exp(-1 *x))

# --- make class ---
class conlayer_left():
    
    def __init__(self,ker,in_c,out_c):
        self.w = tf.Variable(tf.random_normal([ker,ker,in_c,out_c],stddev=0.05))

    def feedforward(self,input,stride=1,dilate=1):
        self.input  = input
        self.layer  = tf.nn.conv2d(input,self.w,strides = [1,stride,stride,1],padding='SAME')
        self.layerA = tf_relu(self.layer)
        return self.layerA

class conlayer_right():
    
    def __init__(self,ker,in_c,out_c):
        self.w = tf.Variable(tf.random_normal([ker,ker,in_c,out_c],stddev=0.05))

    def feedforward(self,input,stride=1,dilate=1,output=1):
        self.input  = input

        current_shape_size = input.shape

        self.layer = tf.nn.conv2d_transpose(input,self.w,
        output_shape=[batch_size] + [int(current_shape_size[1].value*2),int(current_shape_size[2].value*2),int(current_shape_size[3].value/2)],strides=[1,2,2,1],padding='SAME')
        self.layerA = tf_relu(self.layer)
        return self.layerA

# --- make layer ---
# left
l1_1 = conlayer_left(3,3,3)
l1_2 = conlayer_left(3,3,3)
l1_3 = conlayer_left(3,3,3)

l2_1 = conlayer_left(3,3,6)
l2_2 = conlayer_left(3,6,6)
l2_3 = conlayer_left(3,6,6)

l3_1 = conlayer_left(3,6,12)
l3_2 = conlayer_left(3,12,12)
l3_3 = conlayer_left(3,12,12)

l4_1 = conlayer_left(3,12,24)
l4_2 = conlayer_left(3,24,24)
#l4_3 = conlayer_left(3,24,24)
l4_3 = conlayer_left(3,24,12)

'''l5_1 = conlayer_left(3,24,48)
l5_2 = conlayer_left(3,48,48)
l5_3 = conlayer_left(3,48,24)

# right
l6_1 = conlayer_right(3,24,48)
l6_2 = conlayer_left(3,24,24)
l6_3 = conlayer_left(3,24,12)'''

l7_1 = conlayer_right(3,12,24)
l7_2 = conlayer_left(3,12,12)
l7_3 = conlayer_left(3,12,6)

l8_1 = conlayer_right(3,6,12)
l8_2 = conlayer_left(3,6,6)
l8_3 = conlayer_left(3,6,3)

l9_1 = conlayer_right(3,3,6)
l9_2 = conlayer_left(3,3,3)
l9_3 = conlayer_left(3,3,3)

l10_final = conlayer_left(3,3,3)

# ---- make graph ----

layer1_1 = l1_1.feedforward(x)
layer1_2 = l1_2.feedforward(layer1_1)
layer1_3 = l1_3.feedforward(layer1_2)

layer2_Input = tf.nn.max_pool(layer1_3,ksize=[1,2,2,1],strides=[1,2,2,1],padding='VALID')
layer2_1 = l2_1.feedforward(layer2_Input)
layer2_2 = l2_2.feedforward(layer2_1)
layer2_3 = l2_3.feedforward(layer2_2)

layer3_Input = tf.nn.max_pool(layer2_3,ksize=[1,2,2,1],strides=[1,2,2,1],padding='VALID')
layer3_1 = l3_1.feedforward(layer3_Input)
layer3_2 = l3_2.feedforward(layer3_1)
layer3_3 = l3_3.feedforward(layer3_2)

layer4_Input = tf.nn.max_pool(layer3_3,ksize=[1,2,2,1],strides=[1,2,2,1],padding='VALID')
layer4_1 = l4_1.feedforward(layer4_Input)
layer4_2 = l4_2.feedforward(layer4_1)
layer4_3 = l4_3.feedforward(layer4_2)

'''layer5_Input = tf.nn.max_pool(layer4_3,ksize=[1,2,2,1],strides=[1,2,2,1],padding='VALID')
layer5_1 = l5_1.feedforward(layer5_Input)
layer5_2 = l5_2.feedforward(layer5_1)
layer5_3 = l5_3.feedforward(layer5_2)

layer6_Input = tf.concat([layer5_3,layer5_Input],axis=3)
layer6_1 = l6_1.feedforward(layer6_Input)
layer6_2 = l6_2.feedforward(layer6_1)
layer6_3 = l6_3.feedforward(layer6_2)'''

layer4_Input1=tf.reshape(layer4_Input,[-1,layer4_3.shape[1],layer4_3.shape[2],layer4_Input.shape[3]])
layer7_Input = tf.concat([layer4_3,layer4_Input1],axis=3)
layer7_1 = l7_1.feedforward(layer7_Input)
layer7_2 = l7_2.feedforward(layer7_1)
layer7_3 = l7_3.feedforward(layer7_2)

layer3_Input1=tf.reshape(layer3_Input,[-1,layer7_3.shape[1],layer7_3.shape[2],layer3_Input.shape[3]])
layer8_Input = tf.concat([layer7_3,layer3_Input1],axis=3)
layer8_1 = l8_1.feedforward(layer8_Input)
layer8_2 = l8_2.feedforward(layer8_1)
layer8_3 = l8_3.feedforward(layer8_2)

layer2_Input1=tf.reshape(layer2_Input,[-1,layer8_3.shape[1],layer8_3.shape[2],layer2_Input.shape[3]])
layer9_Input = tf.concat([layer8_3,layer2_Input1],axis=3)
layer9_1 = l9_1.feedforward(layer9_Input)
layer9_2 = l9_2.feedforward(layer9_1)
layer9_3 = l9_3.feedforward(layer9_2)

layer10 = l10_final.feedforward(layer9_3)

 #logits = Conv2d(conv1, None, (3, 3), act=False, name='uconv1')
 #y_pred = tf.nn.softmax(logits=logits,dim=3,name='softmax_logits')

#y_pred_cls = tf.argmax(y_pred, dimension=1)

y_pred = tf.nn.softmax(layer10,name='y_pred')
y_pred_cls = tf.argmax(y_pred, dimension=1)

cross_entropy = tf.nn.softmax_cross_entropy_with_logits(logits=layer10,
                                                    labels=y_true)
#cost = tf.reduce_mean(cross_entropy)
cost = tf.reduce_mean(tf.square(layer10-y_true))

optimizer = tf.train.AdamOptimizer(learning_rate=1e-3).minimize(cost)
#correct_prediction = tf.equal(tf.argmax(logits,3),labels)
correct_prediction = tf.equal(y_pred_cls, y_true_cls)
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

#session.run(tf.global_variables_initializer()) # for newer versions
session.run(tf.initialize_all_variables()) # for older versions
train_batch_size = batch_size

### FOR TENSORBOARD TUTORIAL ONLY
#writer= tf.summary.FileWriter('/tmp/tensorboard_tut')
#writer.add_graph(session.graph)

def print_progress(epoch, feed_dict_train, feed_dict_validate, val_loss):
    # Calculate the accuracy on the training-set.
    acc = session.run(accuracy, feed_dict=feed_dict_train)
    val_acc = session.run(accuracy, feed_dict=feed_dict_validate)
    msg = "Epoch {0} --- Training Accuracy: {1:>6.1%}, Validation Accuracy: {2:>6.1%}, Validation Loss: {3:.3f}"
    print(msg.format(epoch + 1, acc, val_acc, val_loss))

   
total_iterations = 0

def optimize(num_iterations):
    # Ensure we update the global variable rather than a local copy.
    global total_iterations

    best_val_loss = float("inf")

    for i in range(total_iterations,
                   total_iterations + num_iterations):

        # Get a batch of training examples.
        # x_batch now holds a batch of images and
        # y_true_batch are the true labels for those images.
        x_batch, y_true_batch = data.train.next_batch(train_batch_size)
        #x_valid_batch, y_valid_batch, _, valid_cls_batch = data.valid.next_batch(train_batch_size)
       
        # Convert shape from [num examples, rows, columns, depth]
        # to [num examples, flattened image shape]

        #x_batch = x_batch.reshape(train_batch_size, img_size_flat)
        #x_valid_batch = x_valid_batch.reshape(train_batch_size, img_size_flat)
        # Put the batch into a dict with the proper names
        # for placeholder variables in the TensorFlow graph.
        length = len(data.valid.labels)
        x1_batch = data.valid.images
        x1_valid_batch = data.valid.labels
        #x1_batch = data.valid.images.reshape(length, img_size_flat)
        #y_true_batch1=tf.reshape(y_true_batch,[y_true_batch.shape[0],64,64,3])

        feed_dict_train = {x: x_batch,
                           y_true: y_true_batch,keep_prob: 0.8}
        
        feed_dict_validate = {x: x1_batch,
                              y_true:x1_valid_batch, keep_prob: 1.0}

        # Run the optimizer using this batch of training data.
        # TensorFlow assigns the variables in feed_dict_train
        # to the placeholder variables and then runs the optimizer.
        session.run(optimizer, feed_dict=feed_dict_train)
        saver = tf.train.Saver()
        saver.save(session, './my_test_model_unet') 

        # Print status at end of each epoch (defined as full pass through training dataset).
        if i % int(data.train.num_examples/batch_size) == 0: 
            val_loss = session.run(cost, feed_dict=feed_dict_validate)
            epoch = int(i / int(data.train.num_examples/batch_size))
            #print(total_iterations + num_iterations, i)            
            print_progress(epoch, feed_dict_train, feed_dict_validate, val_loss)
       

            #print(metrics.confusion_matrix(y_true, y_pred))

    # Update the total number of iterations performed.
    total_iterations += num_iterations
    plt.style.use("ggplot")
    plt.figure()
    plt.plot(plt_epoch,plt_acc)
    plt.plot(plt_epoch,plt_loss)
    plt.xlim([0,epoch])
    plt.ylim([0.0,1.2])
    plt.rcParams['font.size']=10
    plt.title('Accuracy curve of Intima media segmentation')
    plt.xlabel('EPOCH')
    plt.ylabel('ACCURACY/LOSS')
    plt.legend()
    plt.show()
    plt.grid(True)
    
    
optimize(num_iterations=50)
x_batch = data.valid.images
saver = tf.train.import_meta_graph("my_test_model_unet.meta")

graph = tf.get_default_graph()
 
y_pred = graph.get_tensor_by_name("y_pred:0")
# Let's feed the images to the input placeholders
x= graph.get_tensor_by_name("x:0") 
y_true = graph.get_tensor_by_name("y_true:0") 
y_test_images = np.zeros(shape=(100,200,600,3)) 

sess=tf.Session()
feed_dict_testing = {x: x_batch, y_true: y_test_images}
saver.restore(sess, tf.train.latest_checkpoint('./'))
result=sess.run(y_pred, feed_dict=feed_dict_testing)

print(result.shape)
plt.figure()
plt.imshow(result)
plt.axis('off')
plt.title('Generated Mask')
plt.savefig('testing data/'+str(iter)+"c_Generated_Mask.png")
plt.close('all')

**Comment by andresusanopinto** [comment]
This is not a bug about tensorflow_hub. Closing.
